packages/roosterjs-editor-plugins/lib/CustomReplace/CustomReplace.ts

```diff
@@ -0,0 +1,131 @@
+import { Editor, EditorPlugin, cacheGetContentSearcher } from 'roosterjs-editor-core';
+import { PositionType, PluginEvent, PluginEventType } from 'roosterjs-editor-types';
+
+export type Replacement = {
+    sourceString: string;
+    replacementHTML: string;
+    matchSourceCaseSensitive: boolean;
+};
+
+const makeReplacement = (
+    sourceString: string,
+    replacementHTML: string,
+    matchSourceCaseSensitive: boolean
+): Replacement => ({ sourceString, replacementHTML, matchSourceCaseSensitive });
+const defaultReplacements: Replacement[] = [
+    makeReplacement(':)', '🙂', true),
+    makeReplacement(';)', '😉', true),
+    makeReplacement(':O', '😲', true),
+    makeReplacement(':o', '😯', true),
+    makeReplacement('<3', '❤️', true),
+];
+
+/**
+ * Wrapper for CustomReplaceContentEditFeature that provides an API for updating the
+ * content edit feature
+ */
+export default class CustomReplacePlugin implements EditorPlugin {
+    private longestReplacementLength: number;
+    private editor: Editor;
+    private replacements: Replacement[];
+
+    /**
+     * Create instance of CustomReplace plugin
+     * @param features An optional feature set to determine which features the plugin should provide
+     */
+    constructor(replacements: Replacement[] = defaultReplacements) {
+        this.updateReplacements(replacements);
+    }
+
+    /**
+     * Set the replacements that this plugin is looking for.
+     * @param newReplacements new set of replacements for this plugin
+     */
+    updateReplacements(newReplacements: Replacement[]) {
+        this.replacements = newReplacements;
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    /**
+     * Get a friendly name of  this plugin
+     */
+    getName() {
+        return 'CustomReplace';
+    }
+
+    /**
+     * Initialize this plugin
+     * @param editor The editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+        this.editor = null;
+    }
+
+    public onPluginEvent(event: PluginEvent) {
+        if (this.editor.isInIME() || event.eventType != PluginEventType.Input) {
+            return;
+        }
+
+        // Get the matching replacement
+        const range = this.editor.getSelectionRange();
+        if (range == null) {
+            return;
+        }
+        const searcher = cacheGetContentSearcher(event, this.editor);
```

Comment:
Why removed the check of all possible last char? Doing a string search is kind of expensive. We should not do it if we already know the result must be empty.

Reply:
Yes. I wasn't seeing InputEvent.data available in tsc or working in firefox, so I assumed it was not available for a reason.

It [looks like `@types/dom-inputevent` is in the ecosystem and has a definition for InputEvent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/dom-inputevent/index.d.ts
), but there are still issues with using this optimization with inputevent data

- I don't see anywhere in the spec a restriction that InputEvent.data is the character that was inserted by a keyboard event.
- IE11 does not support InputEvent.inputType to distinguish the input events.
- Firefox does not support InputEvent.data _or_ InputEvent.inputType


Because the browser support is so spotty, I saw a risk here for introducing incorrect behavior for basing early exit on this.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/data
![image](https://user-images.githubusercontent.com/1174858/51761728-b7066e00-2082-11e9-9fc8-0bf89af23893.png)

Reply:
We could preserve the behaviour reliably by looking at both key and input events, but we'd have to record when the keyboard event happened and clear it when the current event stack ended, which would mean scheduling a raf (or possibly setTimout -- I'm unsure how timers interact with input event firing).

Via https://unixpapa.com/js/key.html

> To understand the difference between keydown and keypress, it is useful to distinguish between "characters" and "keys". A "key" is a physical button on the computer's keyboard. A "character" is a symbol typed by pressing a button. On a US keyboard, hitting the 4 key while holding down the Shift key typically produces a "dollar sign" character. This is not necessarily the case on every keyboard in the world. In theory, the keydown and keyup events represent keys being pressed or released, while the keypress event represents a character being typed. In practice, this is not always the way it is implemented.
>
>For a while, some browers fired an additional event, called textInput, immediately after keypress. Early versions of the DOM 3 standard intended this as a replacement for the keypress event, but the whole notion was later revoked. Webkit supported this between versions 525 and 533, and I'm told IE supported it, but I never detected that, possibly because Webkit required it to be called "textInput" while IE called it "textinput".
>
>There is also an event called input, supported by all browsers, which is fired just after a change is made to to a textarea or input field. Typically keypress will fire, then the typed charater will appear in the text area, then input will fire. **The input event doesn't actually give any information about what key was typed - you'd have to inspect the textbox to figure it out what changed** - so we don't really consider it a key event and don't really document it here. Though it was originally defined only for textareas and input boxes, I believe there is some movement toward generalizing it to fire on other types of objects as well. 

Which I think is the work being done with InputEvent.data now.

Reply:
That leaves us in a situation where we can only reliably get the inserted character off of a keypress event. If we still make the edit happen on an Input event we're left with a more complicated version of what we had before, since now you'd have to be maintaining state between keypress events and clearing it at the end of the current block of events -- which would require multiple calculations of the cached values on events + setTimeout(0)